File: src/iosrv/system.c
/*
 * (C) Copyright 2018 Intel Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
 * The Government's rights to use, modify, reproduce, release, perform, display,
 * or disclose this software are subject to the terms of the Apache License as
 * provided in Contract No. B609815.
 * Any reproduction of computer software, computer software documentation, or
 * portions thereof marked with this legend must also reproduce the markings.
 */
/**
 * \file
 *
 * System functionalities
 */

#define D_LOGFAC DD_FAC(server)

#include "srv_internal.h"

#define GROUP_ID_MAX_LEN 64
#define ADDR_STR_MAX_LEN 128

#define _S(m) #m
#define S(m) _S(m)

/*
 * Return a URI string, like "ofi+sockets://192.168.1.70:44821", that callers
 * are responsible for freeing with D_FREE.
 */
static char *
create_tag_uri(const char *base_uri, int i)
{
	char   *uri;
	char   *p;
	int	port;

	/* Locate the ":" between the host and the port. */
	p = strrchr(base_uri, ':');
	if (p == NULL)
		return NULL;

	/*
	 * Allocate the tag URI buffer and copy base_uri in except for the
	 * port.
	 */
	D_ALLOC(uri, ADDR_STR_MAX_LEN + 1);
	if (uri == NULL)
		return NULL;
	strncpy(uri, base_uri, p + 1 - base_uri);

	/* Print the port into the tag URI buffer. */
	port = atoi(p + 1) + i;
	D_ASSERT(port <= 65535 /* maximal port number */);
	sprintf(uri + strlen(uri), "%d", port);

	return uri;
}

static int
add_server(crt_group_t *group, d_rank_t rank, char *uri, int ntags)
{
	crt_node_info_t	info;
	int		i;
	int		rc;

	for (i = 0; i < ntags; i++) {
		info.uri = create_tag_uri(uri, i);
		rc = crt_group_node_add(group, rank, i, info);
		D_FREE(info.uri);
		if (rc != 0) {
			D_ERROR("failed to add node: rank=%u tag=%d uri=%s\n",
				rank, i, info.uri);
			return rc;
		}
	}

	return 0;
}

/* This function doesn't try to restore the primary group state upon errors. */
int
dss_sys_map_load(const char *path, crt_group_id_t grpid, d_rank_t self,
		 int ntags)
{
	crt_group_t    *group;
	FILE	       *file;
	char		name[GROUP_ID_MAX_LEN + 1];
	int		size;
	char		all_or_self[5];
	int		i;
	int		rc;

	group = crt_group_lookup(NULL /* primary group */);
	D_ASSERT(group != NULL);

	file = fopen(path, "r");
	if (file == NULL) {
		rc = daos_errno2der(errno);
		D_ERROR("%s: %s\n", path, strerror(errno));
		goto out;
	}

	/* "name daos_server" */
	rc = fscanf(file, "%*s %"S(GROUP_ID_MAX_LEN)"s", name);
	if (rc == EOF) {
		rc = -DER_INVAL;
		goto out_file;
	}
	if (strcmp(name, grpid) != 0) {
		D_ERROR("invalid group name: %s != %s\n", name, grpid);
		rc = -DER_INVAL;
	}

	/* "size 2" */
	rc = fscanf(file, "%*s %d", &size);
	if (rc == EOF) {
		rc = -DER_INVAL;
		goto out_file;
	}

	/* "all" */
	rc = fscanf(file, "%4s", all_or_self);
	if (rc == EOF) {
		rc = -DER_INVAL;
		goto out_file;
	}

	/*
	 * "0 ofi+sockets://10.7.1.70:31416"
	 * "1 ofi+sockets://10.7.1.71:31416"
	 * ...
	 */
	for (i = 0; i < size; i++) {
		d_rank_t	rank;
		char		uri[ADDR_STR_MAX_LEN + 1];

		rc = fscanf(file, "%d %"S(ADDR_STR_MAX_LEN)"s", &rank, uri);
		if (rc == EOF) {
			rc = -DER_INVAL;
			goto out_file;
		}

		if (rank == self) {
			rc = 0;
			continue;
		}

		rc = add_server(group, rank, uri, ntags);
		if (rc != 0) {
			D_ERROR("failed to add server %u %s: %d\n", rank, uri,
				rc);
			goto out_file;
		}
	}

out_file:
	fclose(file);
out:
	return rc;
}

File: src/control/common/proto/system.go
//
// (C) Copyright 2019 Intel Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
// The Government's rights to use, modify, reproduce, release, perform, display,
// or disclose this software are subject to the terms of the Apache License as
// provided in Contract No. 8F-30005.
// Any reproduction of computer software, computer software documentation, or
// portions thereof marked with this legend must also reproduce the markings.
//

package proto

import (
	"net"

	"github.com/pkg/errors"

	ctlpb "github.com/daos-stack/daos/src/control/common/proto/ctl"
	"github.com/daos-stack/daos/src/control/logging"
	"github.com/daos-stack/daos/src/control/system"
)

// MembersToPB converts internal member structs to protobuf equivalents.
func MembersToPB(members system.Members) (pbMembers []*ctlpb.SystemMember, err error) {
	pbMembers = make([]*ctlpb.SystemMember, 0, len(members))

	for i, m := range members {
		if m == nil {
			return nil, errors.Errorf("nil member at index %d", i)
		}

		pbMembers = append(pbMembers, &ctlpb.SystemMember{
			Addr:  m.Addr.String(),
			Uuid:  m.UUID,
			Rank:  m.Rank,
			State: uint32(m.State()),
		})
	}

	return
}

// MembersFromPB converts to member slice from protobuf format.
//
// Don't populate member Addr field if it can't be resolved.
func MembersFromPB(log logging.Logger, pbMembers []*ctlpb.SystemMember) (members system.Members, err error) {
	var addr net.Addr
	members = make(system.Members, 0, len(pbMembers))

	for _, m := range pbMembers {
		addr, err = net.ResolveTCPAddr("tcp", m.Addr)
		if err != nil {
			return
		}

		members = append(members, system.NewMember(m.Rank, m.Uuid, addr,
			system.MemberState(m.State)))
	}

	return
}

// MemberResultsToPB converts system.MemberResults to equivalent protobuf format.
func MemberResultsToPB(results system.MemberResults) []*ctlpb.SystemStopResp_Result {
	pbResults := make([]*ctlpb.SystemStopResp_Result, 0, len(results))

	for _, r := range results {
		pbResult := &ctlpb.SystemStopResp_Result{Rank: r.Rank, Action: r.Action}
		if r.Err != nil {
			pbResult.Errored = true
			pbResult.Msg = r.Err.Error()
		}
		pbResults = append(pbResults, pbResult)
	}

	return pbResults
}

// MemberResultsFromPB converts results from member actions (protobuf format) to
// system.MemberResults.
func MemberResultsFromPB(log logging.Logger, pbResults []*ctlpb.SystemStopResp_Result) system.MemberResults {
	results := make(system.MemberResults, 0, len(pbResults))

	for _, mr := range pbResults {
		var err error
		if mr.Errored {
			err = errors.New(mr.Msg)
		}

		results = append(results, system.NewMemberResult(mr.Rank, mr.Action, err))
	}

	return results
}
